[RDY] CONTRIBUTING.md: Misc improvements

Pyrohh opened the pull request:
Briefly discussed in #2122

[Here is the rendered version.](https://github.com/Pyrohh/neovim/blob/improve-contrib/CONTRIBUTING.md)

The checklist in particular could use some refinement. I've made some changes besides the ones mentioned above, so please offer any advice that comes to mind. Writing is after all much more subjective than code, and I might have been a a _little_ too verbose in certain places.

Commit:
CONTRIBUTING.md: Misc improvements

- In general, many sections have been expanded a lot.
- Defer to the wiki when it makes sense.
- Rewrite a bunch of things to be clearer. There is a lot more
  verbosity, but clarifying as many items as we can in this document is
  preferable to it being done later in the issue tracker.

(alphabetically sorted)
Helped-by: David Granström <david@davidgranstrom.com>
Helped-by: Florian Walch <florian@fwalch.com>
Helped-by: John Szakmeister <john@szakmeister.net>
Helped-by: Justin M. Keyes <justinkz@gmail.com>
Helped-by: Martin Kopischke <martin@kopischke.net>
Helped-by: oni-link <knil.ino@gmail.com>

[ci skip]

oni-link commented on CONTRIBUTING.md:
```diff
@@ -7,100 +7,198 @@
+- Is the code as modular as possible?
+- Can any global/static variables be replaced?
+- Can any function attributes be used?
+- Are variables named intuitively? functions?
```
Is "functions?" at the end a left over? Should it be "Are variables/functions ..."

jszakmeister replied:
It's a fairly common idiom in the English language.  It's an alternate ending to the previous question.

Pyrohh replied:
Indeed it is, but since I did `global/static` variables above, then I'll do as @oni-link suggested for consistency.

oni-link commented on CONTRIBUTING.md:
Should this be "memory leaks/unclosed sockets?"?

fwalch commented on CONTRIBUTING.md:
```diff
@@ -7,100 +7,198 @@
+the following tags prepended to the title:
+
+- `[WIP]` - Work In Progress. The PR will change, so while there is no
+  immediate need for review, although the submitter still might appreciate it.
```
Remove "although"?

Pyrohh replied:
:+1:

jszakmeister commented on CONTRIBUTING.md:
```diff
@@ -7,100 +7,198 @@
+  multiple different builds using different platforms and compilers are
+  used to test your code.
+- The Travis build does not currently run the tests under
+  [valgrind][valgrind], but you are encouraged to do so locally.
```
Is this still true?  I thought this had changed.

Pyrohh replied:
There's `USE_VALGRIND`, but it's undocumented. I guess I'll add it here.

@tarruda just to verify, does it only run under `oldtests`?

tarruda replied:
> edit 2: There's VALGRIND and USE_VALGRIND... that might confuse users.

:+1: 

Yes the variables are different. This patch should fix it:

```diff
diff --git a/src/nvim/testdir/Makefile b/src/nvim/testdir/Makefile
index afa4ce6..1f278de 100644
--- a/src/nvim/testdir/Makefile
+++ b/src/nvim/testdir/Makefile
@@ -42,12 +42,11 @@ SCRIPTS := test_autoformat_join.out                                    \
 SCRIPTS_GUI := test16.out


-ifdef VALGRIND_GDB
-       VGDB := --vgdb=yes     \
-               --vgdb-error=0
-endif
-
-ifdef USE_VALGRIND
+ifdef VALGRIND
+       ifdef GDB
+               VGDB := --vgdb=yes     \
+                                               --vgdb-error=0
+       endif
        VALGRIND_TOOL := --tool=memcheck     \
                         --leak-check=yes    \
                         --track-origins=yes
@@ -61,7 +60,7 @@ ifdef USE_VALGRIND
                         $(VGDB)                                \
                         --trace-children=yes
 else
-       ifdef USE_GDB
+       ifdef GDB
                TOOL = gdb --args
        endif
 endif
```

fwalch commented:
The issues section is targeted towards users that compile neovim themselves, if they use e.g. the PPA, some steps don't apply (`make distclean`) or are different (`:version` might not show commit hash).

General idea: what about changing the section titles to "Reporting problems" and "Submitting contributions" or something similar? Might be clearer what's in the section for someone not that familiar with Github.

jszakmeister commented on CONTRIBUTING.md:
```diff
@@ -7,100 +7,198 @@
+- Travis builds are compiled with the [`-Werror`][gcc-warnings] flag, so
+  if your PR introduces any compiler warnings then the Travis build will fail.
+- If any unit tests fail, the Travis build will fail. They can be run
+  locally with `make test`, which is generally much faster than waiting on
```
We should probably fix this situation, but currently `make test` on runs functional tests, IIRC.  They should run `make unittest` for the unit test suite, and `make functionaltest` (or `make test`) for the functional test suite.  And, I believe we then have *legacy* tests too (`make oldtest`).  All of them are run by Travis, IIRC.

Pyrohh replied:
Well... I guess I was wrong all this time heh. I don't even know the difference between the two (I know the difference between `oldtest` at least).

Pyrohh replied:
> We should probably fix this situation, but currently make test on runs functional tests, IIRC

Would this work?

```patch
diff --git a/Makefile b/Makefile
index 9390463..8eabbb4 100644
--- a/Makefile
+++ b/Makefile
@@ -83,7 +83,7 @@ oldtest: | nvim
 functionaltest: | nvim
 	+$(BUILD_CMD) -C build functionaltest
 
-test: functionaltest
+test: functionaltest unittest
 
 unittest: | nvim
 	+$(BUILD_CMD) -C build unittest
```

fwalch replied:
This would require changes in some `.ci/` files to not run the unit tests twice (they use `make test` instead of `make functionaltest`).

jszakmeister commented on CONTRIBUTING.md:
```diff
@@ -7,100 +7,198 @@
-      place for those comments to live.
+- If necessary, Include further description after a blank line.
+    - Don't make the description too verbose by including obvious things, but
+      don't spare clarifications for anything that could be not so obvious.
```
s/could/may/?

Pyrohh replied:
:+1:

davidgranstrom commented on CONTRIBUTING.md:
```diff
@@ -7,100 +7,198 @@
+  needed, etc.
+- Do ***not*** make cosmetic changes to unrelated files in the same pull
+  request. This creates noise, making reviews harder to do. If your text
+  editor that strips all trailing whitespace in a file when you open/save it,
```
Remove "that"?

Pyrohh replied:
:+1:

jszakmeister commented on CONTRIBUTING.md:
I find the example checklist in SmeatBear's [Best Kept Secrets of Peer Code Review](http://smartbear.com/SmartBear/media/pdfs/best-kept-secrets-of-peer-code-review.pdf) (on pg. 122) useful too:

1. Documentation: All subroutines are commented in clear language.
2. Documentation: Describe what happens with corner-case input.
3. Documentation: Complex algorithms are explained and justified.
4. Documentation: Code that depends on non-obvious behavior in external libraries is documented with reference to external documentation.
5. Documentation: Units of measurement are documented for numeric values.
6. Documentation: Incomplete code is indicated with appropriate distinctive markers (e.g. `TODO` or `FIXME`).
7. Documentation: User-facing documentation is updated (online help, contextual help, tool-tips, version history).
8. Testing: Unit tests are added for new code paths or behaviors.
9. Testing: Unit tests cover errors and invalid parameter cases.
10. Testing: Unit tests demonstrate the algorithm is performing as documented.
11. Testing: Possible null pointers always checked before use.
12. Testing: Array indexes checked to avoid out-of-bound errors.
13. Testing: Don't write new code that is already implemented in an existing, tested API.
14. Testing: New code fixes/implements the issue in question.
15. Error Handling: Invalid parameter values are handled properly early in the subroutine.
16. Error Handling: Error values of null pointers from subroutine invocations are checked.
17. Error Handling: Error handlers clean up state and resources no matter where an error occurs.
18. Error Handling: Memory is released, resources are closed, and reference counters are managed under both error and nonerror conditions.
19. Thread Safety: Global variables are protected by locks or locking subroutines.
20. Thread Safety: Objects accessed by multiple threads are accessed only through a lock.
21. Thread Safety: Locks must be acquired and released in the right order to prevent deadlocks, even in error-handling code.
22. Performance: Objects are duplicated only when necessary.
23. Performance: No busy-wait loops instead of proper thread-synchronization methods.
24. Performance: Memory usage is acceptable even with large inputs.
25. Performance: Optimization that makes code harder to read should only be implemented if a profiler or other tool has indicated that the routine stands to gain from optimization.  These kinds of optimizations should be well-documented and code that performs the same task simply should be preserved somewhere.

Pyrohh commented on CONTRIBUTING.md:
I think `resource leaks works`.

Pyrohh commented on CONTRIBUTING.md:
These are great, I'll add them soon.

Pyrohh commented:
@fwalch 

> The issues section is targeted towards users that compile neovim themselves, if they use e.g. the PPA, some steps don't apply (make distclean) or are different (:version might not show commit hash).

Does the commit hash not show up on release builds or something? If so, why is that the case? I would imagine that only compile flags should be changed by the build type.

I've mentioned `:version`, but what do you recommend in terms of troubleshooting for those not building from source? All I can think of are `brew doctor` currently.

> General idea: what about changing the section titles to "Reporting problems" and "Submitting contributions" or something similar? Might be clearer what's in the section for someone not that familiar with Github.

Good idea, done.

@tarruda at [this wiki page](https://github.com/neovim/neovim/wiki/Building-Neovim), under `Integration tests`, there are still instructions for installing `lupa`, is this still required? I saw a commit saying it isn't just wanted to verify.

tarruda commented:
> @tarruda at this wiki page, under Integration tests, there are still instructions for installing lupa, is this still required? I saw a commit saying it isn't just wanted to verify.

@Pyrohh no, that is no longer necessary.

Pyrohh commented:
I have added a link to the Google style guide for Vimscript (#1475).

@ZyX-I I saw you expressed criticism to that style guide in the aforementioned issue, do you have any other issues with it besides their reasoning to not use `:echoerr`? We should probably just fork it, as the few Vimscript contributions we've gotten usually have a bunch of things wrong with them initially (in terms of style at least).

fwalch commented:
@Pyroh: for the PPA builds, the .git directory is stripped away before building by the build service, so the commit hash is not available (at least not how we currently detect it. Don't know if there's another way).
If a user has installed Neovim from a distribution package, the version of this package might have the relevant information.

jszakmeister commented:
> Does the commit hash not show up on release builds or something? If so, why is that the case? I would imagine that only compile flags should be changed by the build type.

What if you build from a release tarball?  We don't do them just yet, but I expect that we will. :-)

fwalch commented:
For release tarballs, we'll have a meaningful version number instead, so output of `:version` should again be useful. I'll try to see if the commit hash can somehow be passed for the unstable/master-branch PPA builds.

Pyrohh commented:
I re-pushed with many improvements.

justinmk commented:
Under "Reporting problems" let's mention something like this:

    For bug reports, verify whether the bug is present in Vim (not Neovim), and report the Vim version. It's fine to report Vim bugs on the Neovim bug tracker, but it saves everyone time if we know from the start that the bug is not a regression caused by Neovim.

Pyrohh commented:
> Under "Reporting problems" let's mention something like this:

:+1: I've added that.

justinmk commented:
Without reading the changes, I'll mention one thing I was worried about was dumping *too much* information on potential bug reporters / contributors. A long CONTRIBUTING.md could be discouraging if it appears too bureaucratic.

From our recent discussions, I know (and agree) that currently there are too many unanswered questions for new contributors.

To address both of those tensions, it would be ideal if CONTRIBUTING.md contained *only* essential information and pointed to relevant wiki sections for further details.

Pyrohh commented:
@justinmk I think that's a very valid concern. I think most if not all of the "reviewing" stuff could be moved to a wiki page, and some from the "reporting issues" section could too.

kopischke commented:
> When I said reviewing stuff, I just meant the checklist could be moved, the top part may be very helpful for those who want to be reviewers IMO.

:+1:

Pyrohh commented:
I moved the checklist from here to the wiki, as well as removed the link to the VimL style guide for now (we use different indentation requirements).

Pyrohh commented:
Since I've got the core of this done, I've set it to `[RFC]`.

Pyrohh commented:
Does anyone actually use [waffle.io](https://waffle.io/neovim/neovim)? I see it's not even mentioned in our README.

Pyrohh commented:
Perhaps the change to `[RFC]` was too hasty, because I went over it again and slimmed it down a reasonable amount. As far as I'm concerned the main issues have been addressed, but please say if you think otherwise.

Here's the core of what was done recently, as to ease review:
- more brevity in general
- moved build troubleshooting stuff to the wiki
- moved the reviewer checklist from here to the wiki
- removed the link to the VimL style guide for now (we use different indentation requirements).

Pyrohh commented:
@kopischke @fwalch @justinmk Any advice?

fwalch commented on CONTRIBUTING.md:
```diff
@@ -3,104 +3,192 @@
+  multiple different builds using different platforms and compilers are
+  used to test your code.
+- The Travis build does not currently run the tests under
+  [valgrind][valgrind], but you're encouraged to do so locally.
```
Travis does run Valgrind (GCC/Linux build). To do so locally, you can do something like `VALGRIND=1 make test`.

Pyrohh replied:
:+1: will update

kopischke commented on CONTRIBUTING.md:
```diff
@@ -3,104 +3,192 @@
+- If not already running the latest version of Neovim, update to it to see if
+  your problem persists.
+- If you're experiencing compile or runtime warnings/failures, try searching for
+  for the error message(s) you received (if any) on [Neovim's issue tracker][github-issues].
```
Duplicate “for” (before and after lien break)

Pyrohh replied:
:+1:

kopischke commented on CONTRIBUTING.md:
```diff
@@ -3,104 +3,192 @@
+  for the error message(s) you received (if any) on [Neovim's issue tracker][github-issues].
+    - For build issues, see [Building Neovim][wiki-building-troubleshooting].
+    - For runtime issues: try to reproduce them by running `nvim` with the
+      smallest possible `nvimrc` (or none at all via `nvim -u NONE`), as to
```
Either “so as to”, or, preferably IMO, just “to” (no “as”).

Pyrohh replied:
`so as to` works, `to` after a comma sounds odd.

kopischke commented:
Looks pretty good to me, though I think the section about using a review checklist could be more assertive.

Pyrohh commented:
Thanks for the reviews, but I dropped the `Review` commits as I was getting merge conflicts (sometimes 2-3 in a row) every time I changed something in the first commit. I'd like to give credit, but it's just too much of a hassle with the current methodology (at least in this PR, other smaller ones are fine).

> Looks pretty good to me, though I think the section about using a review checklist could be more assertive.

:+1: I moved that to the top of the `Reviewing PRs section`, as it's probably going to be more helpful than being able to review offline (at least in most cases).

kopischke commented:
LGTM (and don’t worry about the review commits; that was very nice of you, but it’s anything but necessary).

Pyrohh commented:
I fixed the [nit about unit tests vs. tests](https://github.com/neovim/neovim/pull/2124#discussion_r26107174) and repushed, is there anything  else needed?

Pyrohh commented:
Added a link to the new `Installing-Neovim#troubleshooting` subpage, it can be expanded on later (it currently just links to [neovim/homebrew-neovim#troubleshooting](https://github.com/neovim/homebrew-neovim#troubleshooting))

oni-link commented on CONTRIBUTING.md:
```diff
@@ -3,104 +3,190 @@
 [waffle]: https://waffle.io/neovim/neovim
-[wiki]: https://github.com/neovim/neovim/wiki/Contributing
+[wiki-building-testing]: https://github.com/neovim/neovim/wiki/Building-Neovim#testing
+[wiki-building-troubleshooting]: https://github.com/neovim/neovim/wiki/Building-Neovim#troubleshootingfaq
```
Should be `https://github.com/neovim/neovim/wiki/Building-Neovim#troubleshooting`

Commit:
CONTRIBUTING.md: Review

Helped-by: Florian Walch <florian@fwalch.com>
Helped-by: oni-link <knil.ino@gmail.com>

Pyrohh commented:
Thanks @oni-link, updated.

Pyrohh commented:
Given all the feedback on this, does anyone have any objections before merging?

fwalch commented on CONTRIBUTING.md:
```diff
@@ -3,104 +3,190 @@
+[VCS][vcs] logs more valuable.
+
+- Try to keep the first line under 72 characters.
+- If necessary, Include further description after a blank line.
```
"Include" -> "include"

Pyrohh replied:
:+1:

fwalch commented on CONTRIBUTING.md:
```diff
@@ -3,104 +3,190 @@
       [For example](https://github.com/neovim/neovim/commits?author=elmart),
-      the following commits are related by task (*Introduce vim namespace*) and
+      the following commits are related by task (*Introduce nvim namespace*) and
       scope (*Contrib YCM*).
```
Should it be "subtask" instead of "scope"?

Pyrohh replied:
:+1:

fwalch commented on CONTRIBUTING.md:
```diff
@@ -3,104 +3,190 @@
+      Some commit messages are pages long, and that's fine if there's no
+      better place for those comments to live.
     - **Recommended:** Prefix logically-related commits with a consistent
-      identifier at the beginning of each commit message.
```
Add "For already used identifiers, have a look a the commit history of the respective file/code parts" or something like that?

Pyrohh replied:
Sure

Pyrohh commented:
@fwalch I amended the review commit, hopefully that's everything :)

fwalch commented:
@Pyrohh :+1: LGTM

Pyrohh commented:
yay
